### DISCUSSION TITLE
날짜 필터링에 사용할 VO를 만드는건 어떨까요?

### DISCUSSION THREAD

**Discussion by kmw2378** [KakaoFunding/back-end#179] [Ideas]
제가 이번 스프린트에 맡은 기능은 아래와 같습니다.
- 주문 내역 조회
- 내가 기여한 펀딩 내역 조회

두 기능 모두 **날짜 필터링**이 존재합니다. 날짜 필터링시 파라미터에 대해 아래 검증 로직이 필요합니다.
- 조회 시작 날짜 < 조회 종료 날짜
- 조회 시작 ~ 종료일 < 1년
- 조회 시작, 종료일 모두 `null` 이여선 안됨

도메인이 다르다보니 Service 마다 파라미터에 대해 검증 메서드를 각각 작성하여 중복 코드가 발생하였습니다.
중복 코드를 없애고자 별도의 VO를 도입하려 하는데 어떻게 생각하실까요??

```java
@Getter
public final class Date {
    private static final int MAX_DATE_PERIOD = 1;

    private final LocalDate startDate;
    private final LocalDate endDate;

    public Date(final LocalDate startDate, final LocalDate endDate) {
        validateDateRange(startDate, endDate);
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public LocalDateTime getStartDateTime() {
        return startDate.atStartOfDay();
    }

    public LocalDateTime getEndDateTime() {
        return endDate.atTime(LocalTime.MAX);
    }

    private void validateDateRange(final LocalDate startDate, final LocalDate endDate) {
        if (startDate == null && endDate == null) {
            throw new IllegalArgumentException();
        }

        if (endDate.isBefore(startDate)) {
            throw new IllegalArgumentException();
        }

        if (ChronoUnit.YEARS.between(endDate, startDate) >= MAX_DATE_PERIOD) {
            throw new IllegalArgumentException();
        }
    }
}
```

추가로 `RepositoryUtils`에 아래 메서드를 추가한다면 Repository내에서 날짜 조건문을 간단히 표현할 수 있습니다!
```java
public static BooleanExpression containsExpression(final DateTimePath<LocalDateTime> dateExpression, final Date date) {
    return dateExpression.between(date.getStartDateTime(), date.getEndDateTime());
}
```

**Comment by YeaChan05** [comment]
VO 도입에 대해서는 동의합니다
하지만 가능하면 추상화 가능한 형태면 좋을거 같습니다
VO를 사용하는 클래스에 따라서 실제 구현내용이 많이 달라질 수 있으니까요

**Reply by kmw2378** [reply]
음.. 날짜 필터링을 하는 로직들의 공통적인 메서드를 표현한 클래스라 추상화가 필요하다 생각하진 않습니다..! 
도메인마다 추가적인 로직이 필요하다면 Service레이어에서 `Date` 의 getter, setter를 사용한 추가 로직 작성은 어떻게 생각하실까요?

**Reply by YeaChan05** [reply]
서비스로직에서 구현하게 되면 별도의 VO를 사용하는 의미가 퇴색되는 느낌이라서요
예를들어 지금 필터링에는 두 날짜 사이가 1년을 넘기면 예외를 던지지만 이부분은 서비스 구현부에 따라 다르게되고, 그렇게되면 매번 서비스로직에서 검증하는 로직을 작성해야 하는데 이를 `Date` 구현체에서 구현하는게 SRP에 더 가까울거 같아요

**Reply by kmw2378** [reply]
아 그러면 저 `Date` 클래스를 추상클래스로 두고 필요한 도메인마다 구현 클래스를 만들어서 사용하자는 말씀이실까요??

**Reply by YeaChan05** [reply]
네
그편이 조금더 유동적으로 수정 가능할거 같아요

**Reply by kmw2378** [reply]
네네 감사합니다! 말씀하신 방식대로 수정해보겠습니다!
